### PR TITLE
Runtime lookup clean up, enable for helper-based tail calls

### PR DESCRIPTION
**Pull request by EgorBo**
Follow up to https://github.com/dotnet/runtime/pull/81635

Clean up some dead code and enable runtime lookup expansion for helper-based tail-calls where we previously only used slow fallbacks.

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -1865,37 +1837,8 @@ GenTree* Compiler::impRuntimeLookupToTree(CORINFO_RESOLVED_TOKEN* pResolvedToken
...
-    indir->gtFlags |= GTF_IND_INVARIANT;
-
-    slot                = gtNewLclvNode(slotLclNum, TYP_I_IMPL);
-    GenTree*      asg   = gtNewAssignNode(slot, indir);
-    GenTreeColon* colon = new (this, GT_COLON) GenTreeColon(TYP_VOID, gtNewNothingNode(), asg);
-    GenTreeQmark* qmark = gtNewQmarkNode(TYP_VOID, relop, colon);
-    impAppendTree(qmark, CHECK_SPILL_NONE, impCurStmtDI);
-
-    return gtNewLclvNode(slotLclNum, TYP_I_IMPL);
+    assert(!pRuntimeLookup->testForFixup);
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:1840]: This should be deleted from JIT/EE interface.

**Comment by EgorBo** [comment]
A couple of [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=204870&view=ms.vss-build-web.run-extensions-tab) in coreclr_tests where we have `tail. callvirt`, e.g:

```diff
; Program:GenInterfaceForwardG
@@ -39,11 +42,26 @@ G_M7779_IG01:        ; bbWeight=1, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref,
        mov      rsi, r9
        ; gcrRegs +[rsi]
 						;; size=44 bbWeight=1 PerfScore 10.58
-G_M7779_IG02:        ; bbWeight=1, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref
+G_M7779_IG02:        ; bbWeight=1, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref, isz
+       mov      rdx, qword ptr [rcx+38H]
+       cmp      qword ptr [rdx+10H], 32
+       jle      SHORT G_M7779_IG05
+						;; size=11 bbWeight=1 PerfScore 6.00
+G_M7779_IG03:        ; bbWeight=0.80, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref, isz
+       mov      rdx, qword ptr [rdx+20H]
+       test     rdx, rdx
+       je       SHORT G_M7779_IG05
+						;; size=9 bbWeight=0.80 PerfScore 2.60
+G_M7779_IG04:        ; bbWeight=0.64, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref, isz
+       jmp      SHORT G_M7779_IG06
+						;; size=2 bbWeight=0.64 PerfScore 1.28
+G_M7779_IG05:        ; bbWeight=0.36, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref
        mov      rdx, 0xD1FFAB1E      ; global ptr
        call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
        ; gcr arg pop 0
        mov      rdx, rax
+						;; size=18 bbWeight=0.36 PerfScore 0.54
+G_M7779_IG06:        ; bbWeight=1, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref
        mov      rcx, rsi
        ; gcrRegs +[rcx]
        mov      r8, 0xD1FFAB1E      ; token handle
@@ -67,8 +85,8 @@ G_M7779_IG02:        ; bbWeight=1, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {},
        ; gcr arg pop 0
        mov      rax, gword ptr [rsp+30H]
        ; gcrRegs +[rax]
-						;; size=91 bbWeight=1 PerfScore 11.50
-G_M7779_IG03:        ; bbWeight=1, epilog, nogc, extend
+						;; size=73 bbWeight=1 PerfScore 10.00
+G_M7779_IG07:        ; bbWeight=1, epilog, nogc, extend
        add      rsp, 96
        pop      rbx
        pop      rsi
@@ -76,7 +94,7 @@ G_M7779_IG03:        ; bbWeight=1, epilog, nogc, extend
        ret      
 						;; size=8 bbWeight=1 PerfScore 2.75
 
-; Total bytes of code 143
```

**Comment by jakobbotsch** [comment]
Nice! Any benchmarks?

BTW, doesn't the block layout look kind of odd here:
```diff
+G_M7779_IG03:        ; bbWeight=0.80, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref, isz
+       mov      rdx, qword ptr [rdx+20H]
+       test     rdx, rdx
+       je       SHORT G_M7779_IG05
+						;; size=9 bbWeight=0.80 PerfScore 2.60
+G_M7779_IG04:        ; bbWeight=0.64, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref, isz
+       jmp      SHORT G_M7779_IG06
+						;; size=2 bbWeight=0.64 PerfScore 1.28
+G_M7779_IG05:        ; bbWeight=0.36, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref
        mov      rdx, 0xD1FFAB1E      ; global ptr
        call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
        ; gcr arg pop 0
        mov      rdx, rax
+						;; size=18 bbWeight=0.36 PerfScore 0.54
+G_M7779_IG06:        ; bbWeight=1, gcrefRegs=00C0 {rsi rdi}, byrefRegs=0000 {}, byref
```

Couldn't this be `jne G_M7779_IG06` directly from the hot block instead?

**Comment by EgorBo** [comment]
> Quick benchmark:

Thanks for checking - I was puzzled on how to properly benchmark it 🙂.

> Couldn't this be jne G_M7779_IG06 directly from the hot block instead?

We emit block layout like this:

```csharp
// prevBb(BBJ_NONE):                    [weight: 1.0]
//     ...
//
// nullcheckBb(BBJ_COND):               [weight: 1.0]
//     if (*fastPathValue == null)
//         goto fallbackBb;
//
// fastPathBb(BBJ_ALWAYS):              [weight: 0.8]
//     rtLookupLcl = *fastPathValue;
//     goto block;
//
// fallbackBb(BBJ_NONE):                [weight: 0.2]
//     rtLookupLcl = HelperCall();
//
// block(...):                          [weight: 1.0]
//     use(rtLookupLcl);
```
We probably need to re-order fastPathBb and fallbackBb

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/vm/prestub.cpp [hunk 2]
```diff
@@ -2843,7 +2843,6 @@ void ProcessDynamicDictionaryLookup(TransitionBlock *           pTransitionBlock
 
     TADDR genericContextPtr = *(TADDR*)GetFirstArgumentRegisterValuePtr(pTransitionBlock);
 
-    pResult->testForFixup = pResult->testForNull = false;
```
Comment (AndyAyersMS) [inline src/coreclr/vm/prestub.cpp:2846]: Shouldn't we keep the `pResult->testForNull = false;` part?

Comment (EgorBo) [reply src/coreclr/vm/prestub.cpp:2846]: Good catch, since the caller doesn't zero initialize the struct we need to init all fields

**Pull request merged**